docs/proposals/ktl.md

```diff
@@ -0,0 +1,243 @@
+- fix issues with the existing kubectl commands
+- dog fooding the plugin mechanism for core commands
+
+## Reference-level explanation
+
+- build a new cli binary `ktl` (kubectl2) under the `kubernetes/ktl` repo that dispatches to commands developed in other repos
+- keep old cli commands in `kubernetes/kubernetes/cmd/kubectl` and vendor them into `kubectl/ktl`
+- build new cli commands in `kubernetes/kubectl` and vendor them into `kurnetes/ktl`
+- build common cli Kubernetes client infrastructure and libraries that can be used to develop a decentralized cli ecosystem
+
+### ktl
+
+In a new repo (`kubernetes/ktl`), create the `ktl` binary that dispatches (both statically and dynamically)
+to commands developed in other repos.
+
+#### Dispatch
+
+Static dispatch:
+
+- vendor in kubernetes/kubernetes/cmd/kubectl cobra commands, and register them under `ktl kube`
+- vendor in kubernetes/kubectl/cmd cobra commands, and register them under `ktl`
+- vendor in commands from other sources as needed over time
+
+Dynamic dispatch:
+
+- support git-style plugins that allow registering new commands under `ktl`
+  - use simplified version of the kubectl plugin implementation - make configuration files optional
+- plugins only purpose is discovery of kubernetes related commands
+  - plugins can leverage shared client cli libraries (whether they are installed as plugins or not)
+- by default throw an error if plugin names conflict with existing commands
+  - this is configurable
+- plugins can be disabled through modifying `~/.ktlconfig`
+
+Overriding existing commands:
+
+- support command alias' to allow overriding one command with another
+  - allows plugins to extend and override built-in commands
+  - allows `ktl kube *` commands to be aliased to `ktl *`
+
+#### Configuration
+
+Use [viper](https://github.com/spf13/viper) to configure dispatch precedence
+
+### Cli commands
+
+All new cli commands should be built outside of kubernetes/kubernetes and vendored into the kubernetes/ktl.  Command repos
+implement an interface that returns a list of cobra (sp13) commands, which ktl can register under the root.
+
+Initially new cli commands can be built in the kubernetes/kubectl, but if needed, development maybe moved to other repos.
+
+#### Conventions
+
+- ktl top level commands are one of
+  - command groups that work against a specific API group/versions (e.g. isto, svcat, kube)  e.g. `create`
+    for a given resource.
+  - generic commands that are agnostic to specific APIs, but instead discover API metadata and work against
+    all APIs following published conventions.  e.g. a `scale` that works against anything with a scale sub resource
+
+
+### Library infrastructure
+
+Develop shared set of client / cli libraries that handle registering flags / environment variables / config files
+and provide common functionality to cli clients of Kubernetes.
+
+- reading and parsing kubeconfig
+- printing libraries
```

Comment:
I think we're less interested in the library and more interested in consistent flags and behavior.  I'd really like to see this gotten "right" in `kubectl` before trying again.  I'm not convinced that we've learned from prior mistakes.  `options`, `addflags`, and `toX` provide consistency, flexibility, and ease to the generic apiserver.  Seems like we could transform `kubectl` to use something like that an prove it works.

Reply:
I don't think we should block on perfecting something in kubectl.  The iteration cycle will be much slower.

Reply:
I think the end result of this entire effort will be two CLIs - kubectl (slower) and ktl (faster, some improvements, some differences).  We aren't going to kill kubectl except by having something materially better, so we'll continue to need to evolve kubectl slowly, which means duplication and split effort.  I'm not as worried about the libraries in the scheme of things, although i think printers is something that's on an arc to be correct.